src/config.rs

```diff
@@ -85,23 +106,82 @@ impl LayeredConfigs {
 
     /// Creates new merged config.
     pub fn merge(&self) -> config::Config {
-        let config_sources = [
-            Some(&self.default),
-            Some(&self.env_base),
-            self.user.as_ref(),
-            self.repo.as_ref(),
-            Some(&self.env_overrides),
-            self.arg_overrides.as_ref(),
-        ];
-        config_sources
+        self.sources()
             .into_iter()
-            .flatten()
+            .map(|(_, config)| config)
             .fold(config::Config::builder(), |builder, source| {
                 builder.add_source(source.clone())
             })
             .build()
             .expect("loaded configs should be merged without error")
     }
+
+    fn sources(&self) -> Vec<(ConfigSource, &config::Config)> {
+        let config_sources = [
+            (ConfigSource::Default, Some(&self.default)),
+            (ConfigSource::Env, Some(&self.env_base)),
+            (ConfigSource::User, self.user.as_ref()),
+            (ConfigSource::Repo, self.repo.as_ref()),
+            (ConfigSource::Env, Some(&self.env_overrides)),
+            (ConfigSource::CommandArg, self.arg_overrides.as_ref()),
+        ];
+        config_sources
+            .into_iter()
+            .filter_map(|(source, config)| config.map(|c| (source, c)))
+            .collect_vec()
+    }
+
+    pub fn resolved_config_values(
+        &self,
+        filter_prefix: &[&str],
+    ) -> Result<Vec<AnnotatedValue>, ConfigError> {
+        // Collect annotated values from each config.
+        let mut config_vals = vec![];
+
+        for (source, config) in self.sources() {
+            let config_table = config.collect()?.into();
+            let mut config_stack: Vec<(Vec<&str>, &config::Value)> = vec![(vec![], &config_table)];
+            while let Some((path, value)) = config_stack.pop() {
+                // Skip if path doesn't match filter_prefix.
+                if filter_prefix.iter().zip(path.iter()).any(|(a, b)| *a != *b) {
+                    continue;
```

Comment:
Instead, the initial `config_stack` can be narrowed to the `filter_prefix`.

```rust
filter_prefix.iter().try_fold(config_table, |value, &key| {
                value
                    .into_table()
                    .ok()
                    .and_then(|mut table| table.remove(key))
            })
```

If we could build a `config::path::Expression` safely, it should be just `config.get(path)`, but there's no public API to convert names vec to the path expression.

Reply:
Nope, to initialize the `config_stack = [prefixed_root_entry]` set.

Reply:
But then that's basically copy-pasting the table logic from below, isn't it? If so, I feel like the miniscule performance gain wouldn't be worth the extra code bloat of unrolling the first iteration.

Reply:
Ohhhh, nevermind, you're saying instead of iterating the whole config and filtering I might as well get the value directly from the config, since at the config level it can't have multiple values for the same key anyway. Sounds good, done.